### PR TITLE
Fix ARMv7 generic boot against SMP & pager, support pager in plat-stm

### PR DESCRIPTION
**Pull request by etienne-lms**
A fix pager support for armv7.
A fix for PL310 operations during generic boot.
Enable support of pager TEE on plat-stm.

### CODE REVIEW

File: core/arch/arm/kernel/link.mk [hunk 1]
```diff
@@ -30,6 +30,7 @@ entries-unpaged += sm_init
 entries-unpaged += core_init_mmu_regs
 entries-unpaged += sem_cpu_sync
 entries-unpaged += generic_boot_get_handlers
+entries-unpaged += plat_cpu_reset_late
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/link.mk]: Use the `KEEP_PAGER()` macro instead please. A good spot is probably somewhere in `core/arch/arm/kernel/generic_boot.c`

File: core/arch/arm/plat-stm/platform_config.h [hunk 2]
```diff
@@ -42,6 +42,10 @@
 #define CFG_DDR_TEETZ_RESERVED_START	0x7E000000
 #define CFG_DDR_TEETZ_RESERVED_SIZE	0x01E00000
 #endif
+#ifndef CFG_IRAM_TEETZ_RESERVED_START
+#define CFG_IRAM_TEETZ_RESERVED_START	0x7FE00000
+#define CFG_IRAM_TEETZ_RESERVED_SIZE	0x00040000
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/platform_config.h]: Any reason why this isn't defined to `CFG_CORE_TZSRAM_EMUL_SIZE`?

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 3]
```diff
@@ -218,6 +218,63 @@ END_FUNC reset
...
+	bl	arm_cl2_cleaninvbypa
+#endif
+	ldr	r0, =\vbase
+	ldr	r1, =\vend
+	sub	r1, #1
+	bl	arm_cl1_d_cleaninvbyva
+	.endm
+
+#ifdef CFG_BOOT_SYNC_CPU
+LOCAL_FUNC flush_cpu_semaphores , :
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: ok

File: core/arch/arm/kernel/link.mk (hunk 1 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/link.mk]: ok

Comment (etienne-lms) [inline core/arch/arm/kernel/link.mk]: Actually using `KEEP_PAGER(plat_cpu_reset_late)` fails: `plat_cpu_reset_late` is linked inside the 'int' sections, not the 'unpaged' section. I must find out why.

File: core/arch/arm/kernel/generic_boot.c [hunk 4]
```diff
@@ -225,7 +237,8 @@ static void init_runtime(unsigned long pageable_part)
 	malloc_add_pool(__heap2_start, __heap2_end - __heap2_start);
 
 	hashes = malloc(hash_size);
-	IMSG("Pager is enabled. Hashes: %zu bytes", hash_size);
+	IMSG("Pager is enabled. Hashes: %zu bytes, %zukB pool",
+			hash_size, pageable_size >> 10);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: `pageable_size / 1024` is preferable over `pageable_size >> 10`

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 5]
```diff
@@ -218,6 +176,48 @@ END_FUNC reset
...
 	.endm
 
+	/*
+	 * Cache maintenance during entry: handle outer cache.
+	 * End address is exclusive: first byte not to be changed.
+	 * Note however arm_clX_inv/cleanbyva operate on full cache lines.
+	 */
+	.macro inval_cache_vrange vbase, vend
+#ifdef CFG_PL310
+		/* this asserts vmem range is flat mapped (FIXME: add assertion) */
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Please fix in one way or the other. :-)

File: core/arch/arm/kernel/generic_boot.c (hunk 4 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: ok, but this is buggy. it shows the size of the pageable area, not the number of physical pages available the the pager, which was the value I was expecting.
There is already a way to get this info: from pager statistics, but it requires `CFG_WITH_STATS=y`. I think i will discard this change.

File: core/arch/arm/kernel/generic_entry_a32.S (hunk 5 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: ok...

File: core/arch/arm/kernel/generic_boot.c (hunk 4 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: OK

**Comment by etienne-lms** [comment]
2 remaining comments addressed, i hope the right way...

**Comment by jenswi-linaro** [comment]
Nice cleanup in `generic_entry_a32.S`
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 6]
```diff
@@ -218,6 +193,63 @@ END_FUNC reset
...
+		/* check va=pa for base address of target area */
+		bl	cpu_mmu_enabled
+		cmp	r0, #0
+		beq	1f
+		ldr	r0, =\vbase
+		bl	virt_to_phys
+		ldr	r1, =\vbase
+		cmp	r0, r1
+		ldrne	r0, =__LINE__
+		blne	panic_at_line
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: i feel this is a bit ugly. I will shrink the 'va=pa' test into a single macro for both cache inval and flush.

**Comment by etienne-lms** [comment]
sorry this is not squashed yet, but I  experience issues with pager and core-debug mode enable together.
i want to check whether it is my platform or if there is a generic issue.

**Comment by etienne-lms** [comment]
`Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (b2260)`

issues in pager and debug will be tracked from specific P-Rs/Issues.

Jens, i've modified the 'flat map assertion' in generic_entry_a32.S since you tagged the P-R. can you crosscheck you're fine with this ?

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 7]
```diff
@@ -50,6 +50,51 @@ gdb_bootargs:
...
+	.asciz __FILE__
+
+/*
+ * void assert_flat_mapped_range(uint32_t vaddr, uint32_t line)
+ */
+.section .text.init
+LOCAL_FUNC __assert_flat_mapped_range , :
+UNWIND(	.fnstart)
+UNWIND(	.cantunwind)
+	push	{ r4-r5, lr }
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Please push an even number of registers to maintain required alignment of the stack pointer.

**Comment by jenswi-linaro** [comment]
With my comment addressed I'm OK with this PR again.

File: core/arch/arm/kernel/generic_entry_a32.S (hunk 7 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Looks like an odd number to me.

**Comment by jenswi-linaro** [comment]
Looks good.

**Comment by jforissier** [comment]
@etienne-lms I understand you can squash/rebase and apply Jens' R-b. Thanks!

**Comment by etienne-lms** [comment]
rebase required...

**Pull request merged**